File: bond/PHP/prelude.php
/// python-bond PHP interface setup
/// NOTE: use /// for comments only, as this code is transformed into a single
///       line to be injected into the interpreter *without parsing*.

/// Redirect normal output
$__PY_BOND_BUFFERS = array(
    "STDOUT" => "",
    "STDERR" => ""
);

class __PY_BOND_BUFFERED
{
  public $name;

  public function stream_open($path, $mode, $options, &$opened_path)
  {
    global $__PY_BOND_BUFFERS;
    $path = strtoupper(substr(strstr($path, "://"), 3));
    if(!isset($__PY_BOND_BUFFERS[$path]))
      return false;
    $this->name = $path;
    return true;
  }

  public function stream_write($data)
  {
    global $__PY_BOND_BUFFERS;
    $buffer = &$__PY_BOND_BUFFERS[$this->name];
    $buffer .= $data;
    return strlen($data);
  }
}


/// Redefine standard streams
$__PY_BOND_CHANNELS = array(
    "STDIN" => fopen("php://stdin", "r"),
    "STDOUT" => fopen("php://stdout", "w"),
    "STDERR" => fopen("php://stderr", "w")
);

stream_wrapper_unregister("php");
stream_wrapper_register("php", "__PY_BOND_BUFFERED");

if(!defined("STDIN"))
  define('STDIN', null);
if(!defined("STDOUT"))
  define('STDOUT', fopen("php://stdout", "w"));
if(!defined("STDERR"))
  define('STDERR', fopen("php://stderr", "w"));


/// Define our own i/o methods
function __PY_BOND_output($buffer, $phase)
{
  fwrite(STDOUT, $buffer);
}

function __PY_BOND_getline()
{
  global $__PY_BOND_CHANNELS;
  return rtrim(fgets($__PY_BOND_CHANNELS['STDIN']));
}

function __PY_BOND_sendline($line = '')
{
  global $__PY_BOND_CHANNELS;
  $stdout = $__PY_BOND_CHANNELS['STDOUT'];
  fwrite($stdout, $line . "\n");
  fflush($stdout);
}


/// Serialization methods
class __PY_BOND_SerializationException extends Exception {}

function __PY_BOND_dumps($data)
{
  $code = json_encode($data);
  if(json_last_error())
    throw new __PY_BOND_SerializationException(@"cannot encode $data");
  return $code;
}

function __PY_BOND_loads($string)
{
  return json_decode($string);
}


/// some utilities to get/reset the error state
function __PY_BOND_clear_error()
{
  set_error_handler(null, 0);
  @trigger_error(null);
  restore_error_handler();
}

function __PY_BOND_get_error()
{
  $err = error_get_last();
  if($err) $err = $err["message"];
  return $err;
}


/// Recursive repl
$__PY_BOND_TRANS_EXCEPT = null;

function __PY_BOND_call($name, $args)
{
  $code = __PY_BOND_dumps(array($name, $args));
  __PY_BOND_sendline("CALL $code");
  return __PY_BOND_repl();
}

function __PY_BOND_repl()
{
  global $__PY_BOND_BUFFERS, $__PY_BOND_TRANS_EXCEPT;

  while($line = __PY_BOND_getline())
  {
    $line = explode(" ", $line, 2);
    $cmd = $line[0];
    $args = (count($line) > 1? __PY_BOND_loads($line[1]): array());

    $ret = null;
    $err = null;
    switch($cmd)
    {
    case "EVAL":
      try
      {
	__PY_BOND_clear_error();
	$ret = @eval("return $args;");
	$err = __PY_BOND_get_error();
      }
      catch(Exception $e)
      {
	$err = $e;
      }
      break;

    case "EVAL_BLOCK":
      try
      {
	__PY_BOND_clear_error();
	@eval($args);
	$err = __PY_BOND_get_error();
      }
      catch(Exception $e)
      {
	$err = $e;
      }
      break;

    case "EXPORT":
      $name = $args;
      if(function_exists($name))
	$err = "Function \"$name\" already exists";
      else
      {
	$code = "function $name() { return __PY_BOND_call('$args', func_get_args()); }";
	__PY_BOND_clear_error();
	@eval($code);
	$err = __PY_BOND_get_error();
      }
      break;

    case "CALL":
      try
      {
	$name = $args[0];

	__PY_BOND_clear_error();
	if(preg_match("/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/", $name) || is_callable($name))
	{
	  /// special-case regular functions to avoid fatal errors in PHP
	  $ret = @call_user_func_array($args[0], $args[1]);
	}
	else
	{
	  /// construct a string that we can interpret "function-like", to
	  /// handle also function references and method calls uniformly
	  $args_ = array();
	  foreach($args[1] as $el)
	    $args_[] = var_export($el, true);
	  $args_ = implode(", ", $args_);
	  $ret = @eval("return " . $name . "(" . $args_ . ");");
	}
	$err = __PY_BOND_get_error();
      }
      catch(Exception $e)
      {
	$err = $e;
      }
      break;

    case "RETURN":
      return $args;

    case "EXCEPT":
      throw new Exception($args);

    case "ERROR":
      throw new __PY_BOND_SerializationException($args);

    default:
      exit(1);
    }

    /// redirected channels
    ob_flush();
    foreach($__PY_BOND_BUFFERS as $chan => &$buf)
    {
      if(strlen($buf))
      {
	$code = __PY_BOND_dumps(array($chan, $buf));
	__PY_BOND_sendline("OUTPUT $code");
	$buf = "";
      }
    }

    /// error state
    $state = "RETURN";
    if($err)
    {
      if($err instanceOf __PY_BOND_SerializationException)
      {
	$state = "ERROR";
	$ret = $err->getMessage();
      }
      else
      {
	$state = "EXCEPT";
	if($err instanceOf Exception)
	  $ret = ($__PY_BOND_TRANS_EXCEPT? $err: $err->getMessage());
	else
	  $ret = @"$err";
      }
    }
    $code = null;
    try
    {
      $code = __PY_BOND_dumps($ret);
    }
    catch(Exception $e)
    {
      $state = "ERROR";
      $code = __PY_BOND_dumps($e->getMessage());
    }
    __PY_BOND_sendline("$state $code");
  }
  return 0;
}

function __PY_BOND_start($trans_except)
{
  global $__PY_BOND_TRANS_EXCEPT;
  ob_start('__PY_BOND_output');
  $__PY_BOND_TRANS_EXCEPT = (bool)($trans_except);
  __PY_BOND_sendline("READY");
  $ret = __PY_BOND_repl();
  __PY_BOND_sendline("BYE");
  exit($ret);
}

File: bond/Python/prelude.py
# python-bond Python interface setup
import base64
import cPickle
import cStringIO
import os
import sys


# Redirect normal output
__PY_BOND_BUFFERS = {
    "STDOUT": cStringIO.StringIO(),
    "STDERR": cStringIO.StringIO()
}

__PY_BOND_CHANNELS = {
    "STDIN": sys.stdin,
    "STDOUT": sys.stdout,
    "STDERR": sys.stderr
}


# Define our own i/o methods
def __PY_BOND_getline():
    return __PY_BOND_CHANNELS['STDIN'].readline().rstrip()

def __PY_BOND_sendline(line=""):
    stdout = __PY_BOND_CHANNELS['STDOUT']
    stdout.write(line + "\n")
    stdout.flush()


# Serialization methods
class __PY_BOND_SerializationException(cPickle.PicklingError):
    pass

__PY_BOND_PROTOCOL = None

def __PY_BOND_dumps(*args):
    try:
        ret = base64.b64encode(cPickle.dumps(args, __PY_BOND_PROTOCOL))
    except cPickle.PicklingError:
        raise __PY_BOND_SerializationException("cannot encode {data}".format(data=str(args)))
    return ret

def __PY_BOND_loads(string):
    return cPickle.loads(base64.b64decode(string))[0]


# Recursive repl
__PY_BOND_TRANS_EXCEPT = None

def __PY_BOND_call(name, args):
    code = __PY_BOND_dumps([name, args])
    __PY_BOND_sendline("CALL {code}".format(code=code))
    return __PY_BOND_repl()

def __PY_BOND_export(name):
    globals()[name] = lambda *args: __PY_BOND_call(name, args)

def __PY_BOND_repl():
    while True:
        line = __PY_BOND_getline()
        if len(line) == 0:
            break

        line = line.split(' ', 1)
        cmd = str(line[0])
        args = __PY_BOND_loads(line[1]) if len(line) > 1 else []

        ret = None
        err = None
        if cmd == "EVAL" or cmd == "EVAL_BLOCK":
            try:
                mode = 'eval' if cmd == "EVAL" else 'exec'
                ret = eval(compile(args, '<string>', mode), globals())
            except Exception as e:
                err = e

        elif cmd == "EXPORT":
            __PY_BOND_export(args)

        elif cmd == "CALL":
            try:
                func = eval(compile(args[0], '<string>', 'eval'), globals())
                ret = func(*args[1])
            except Exception as e:
                err = e

        elif cmd == "RETURN":
            return args

        elif cmd == "EXCEPT":
            raise args

        elif cmd == "ERROR":
            raise __PY_BOND_SerializationException(args)

        else:
            exit(1)

        # redirected channels
        for chan, buf in __PY_BOND_BUFFERS.iteritems():
            if buf.tell():
                output = buf.getvalue()
                code = __PY_BOND_dumps([chan, output])
                __PY_BOND_sendline("OUTPUT {code}".format(code=code))
                buf.truncate(0)

        # error state
        state = "RETURN"
        if err is not None:
            if isinstance(err, __PY_BOND_SerializationException):
                state = "ERROR"
                ret = str(err)
            else:
                state = "EXCEPT"
                ret = err if __PY_BOND_TRANS_EXCEPT else str(err)
        try:
            code = __PY_BOND_dumps(ret)
        except Exception as e:
            state = "ERROR"
            code = __PY_BOND_dumps(str(e))
        __PY_BOND_sendline("{state} {code}".format(state=state, code=code))

    # stream ended
    return 0


def __PY_BOND_start(trans_except, protocol):
    global __PY_BOND_TRANS_EXCEPT, __PY_BOND_PROTOCOL

    sys.stdout = __PY_BOND_BUFFERS['STDOUT']
    sys.stderr = __PY_BOND_BUFFERS['STDERR']
    sys.stdin = open(os.devnull)

    __PY_BOND_TRANS_EXCEPT = trans_except
    __PY_BOND_PROTOCOL = protocol
    __PY_BOND_sendline("READY")
    ret = __PY_BOND_repl()
    __PY_BOND_sendline("BYE")
    exit(ret)
